### ISSUE TITLE
Security Issue

### ISSUE THREAD

**Issue by pizzeys** [swiftmailer/swiftmailer#844]
There's a bit of noise going around at the moment regarding a vulnerability in PHPMailer where if an attacker can control the from address a script uses they can pass arbitrary arguments to the sendmail binary, which, if it's actually sendmail, results in being able to write to an arbitrary file (and so on most PHP setups, likely RCE) and if not at least allows to pass other parameters the developer perhaps didn't intend.

Swiftmailer is actually vulnerable to the same bug, if using the Swift_MailTransport transport - for example, one can do:

setFrom(array('"x@y.com\" -X/tmp/test"@test.php' => 'John Doe'))

Is this expected? It would be quite easy for a script to be using this in a dangerous way, so it should probably be made clear in the documentation that this parameter should be handled with kid gloves, if so.

**Comment by fabpot** [comment]
Where you able to reproduce this with Swiftmailer? I didn't, so if you did, can you tell me how (email is probably better)? Thanks.

**Comment by joserprieto** [comment]
The complete vulnerability is described here:

http://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10033-Vuln.html

And the key concept is:

> Further research however shown that the validation is done using the RFC 3696 specification.
> 
> The RFC allows emails to contain spaces when quoted with ". 
> According to the spec, the following email address is valid:
> 
> " some test email address with spaces"@weird-email.com
> 

I can't test this right now, but, if you can send an email like this:

```
$email_from = '"attacker\" -oQ/tmp/ -X/var/www/cache/phpcode.php  some"@email.com';
$msg_body  = "<?php phpinfo(); ?>";
```

You can test the vulnerability.

Regards.

**Comment by pizzeys** [comment]
It looks like this may have previously been fixed here:

https://github.com/swiftmailer/swiftmailer/issues/520

But then reverted due to 'mail() in PHP already escapes this argument.' However that's not the case - PHP passes the parameters through escapeshellcmd() but that doesn't prevent the appending of additional arguments, which is the issue here.

However, that may mean that the fix used there is appropriate for re-use here perhaps?

**Comment by MartinZubek** [comment]
So how does it look like guys? Should we be worried?

**Comment by z38** [comment]
Note that the first fix for CVE-2016-10033 by PHPMailer was incomplete. It seems one has to be quite careful when escaping the additional parameters (CVE-2016-10045):

http://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10045-Vuln-Patch-Bypass.html

> The patch for CVE-2016-10033 vulnerability added in PHPMailer 5.2.17 
> sanitizes the $Sender variable by applying escapeshellarg() escaping
> before the value is passed to mail() function.
> 
> It does not however take into account the clashing of the 
> escapeshellarg() function with internal escaping with escapeshellcmd()
> performed by mail() function on the 5th parameter.
> 
> As a result it is possible to inject an extra quote that does not get
> properly escaped and break out of the escapeshellarg() protection 
> applied by the patch in PHPMailer 5.2.17.

As mentioned, PHP escapes the fifth argument with `escapeshellcmd`: [ext/standard/mail.c](https://github.com/php/php-src/blob/php-7.0.14/ext/standard/mail.c#L171)

**Comment by EndeavourAccuracy** [comment]
> So how does it look like guys? Should we be worried?

The vulnerability [description](http://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10033-Vuln.html) mentions [RFC 3696](https://tools.ietf.org/html/rfc3696), which includes:

> In addition to quoting using the backslash character, conventional double-quote characters may be used to surround strings.

However, at least until 7.1.0 (that has a [ChangeLog](http://php.net/ChangeLog-7.php) which includes: "Implemented email validation as per RFC 6531."), the [description](http://php.net/manual/en/filter.filters.validate.php) of PHP's own FILTER_VALIDATE_EMAIL applies:

> In general, this validates e-mail addresses against the syntax in RFC 822, [...]

That [RFC 822](https://www.ietf.org/rfc/rfc0822.txt), as someone also mentioned [here](http://php.net/manual/en/function.filter-var.php#112492), does not accept double-quote characters:

> I found some addresses that FILTER_VALIDATE_EMAIL rejects, but RFC5321 permits: [...] '"much.more unusual"@example.com'"

This means you should be safe if you validate (and only accept) e-mail addresses that pass...
`if (filter_var ($sEmail, FILTER_VALIDATE_EMAIL) !== FALSE)`
...which is something you should be doing anyway.

It would be crazy to accept and store an e-mail address such as...
`$sEmail = '"attacker\" -oQ/tmp/ -X/var/www/cache/phpcode.php  some"@email.com';`
..and merely rely on other people's libraries to keep you(r server) safe.

The patch bypass [code](http://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10045-Vuln-Patch-Bypass.html) includes `\"` as example user input, and accepting such input would be even more amateurish. Before you do anything with strings, use htmlspecialchars() or something similar:
`htmlspecialchars ($sEmail, ENT_QUOTES);`

You should not have to (constantly) check if your PHPMailer or Swift Mailer needs updating because vulnerabilities in their code were detected. Fix your own code instead, if necessary.

**Comment by EndeavourAccuracy** [comment]
> So how does it look like guys? Should we be worried?

The vulnerability [description](http://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10033-Vuln.html) mentions [RFC 3696](https://tools.ietf.org/html/rfc3696), which includes:

> In addition to quoting using the backslash character, conventional double-quote characters may be used to surround strings.

However, at least until 7.1.0 (that has a [ChangeLog](http://php.net/ChangeLog-7.php) which includes: "Implemented email validation as per RFC 6531."), the [description](http://php.net/manual/en/filter.filters.validate.php) of PHP's own FILTER_VALIDATE_EMAIL applies:

> In general, this validates e-mail addresses against the syntax in RFC 822, [...]

Ignore [RFC 822](https://www.ietf.org/rfc/rfc0822.txt) though. As someone also mentioned [here](http://php.net/manual/en/function.filter-var.php#112492), FILTER_VALIDATE_EMAIL does not accept double-quote characters:

> I found some addresses that FILTER_VALIDATE_EMAIL rejects, but RFC5321 permits: [...] '"much.more unusual"@example.com'"

This means you should be safe if you validate (and only accept) e-mail addresses that pass...
`if (filter_var ($sEmail, FILTER_VALIDATE_EMAIL) !== FALSE)`
...which is something you should be doing anyway.

It would be crazy to accept and store an e-mail address such as...
`$sEmail = '"attacker\" -oQ/tmp/ -X/var/www/cache/phpcode.php  some"@email.com';`
..and merely rely on other people's libraries to keep you(r server) safe.

The patch bypass [code](http://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10045-Vuln-Patch-Bypass.html) includes `\"` as example user input, and accepting such input would be even more amateurish. Before you do anything with strings, use htmlspecialchars() or something similar:
`htmlspecialchars ($sEmail, ENT_QUOTES);`

You should not have to (constantly) check if your PHPMailer or Swift Mailer needs updating because vulnerabilities in their code were detected. Fix your own code instead, if necessary.

**Comment by mturalenka** [comment]
@fabpot in case of found CVE-2016-10033, Phmailer uses same mail() function, so how does this reverting protect from vulnerability?
i wrote simple code according to https://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10033-Vuln.html and modified a bit $email_from to pass RFC 2822
```
<?php
require_once 'lib/swift_required.php';

// Create the Transport
$transport = Swift_MailTransport::newInstance();

// Create the Mailer using your created Transport
$mailer = Swift_Mailer::newInstance($transport);

$email_from = '"Attacker@test.com\" -X/var/www/test_swiftmailer/phpcode.php -oQ/tmp test"@test.com';
$msg_body  = "<?php phpinfo(); ?>";
// Create a message
$message = Swift_Message::newInstance('Wonderful Subject')
                        ->setFrom($email_from)
                        ->setTo('mturalenka@scnsoft.com')
                        ->setBody($msg_body)
;

// Send the message
$result = $mailer->send($message);
```
and `phpcode.php` file was created and accessible with msg_body inside.
used PHP 5.6.12-1+deb.sury.org~trusty+1

**Comment by Synchro** [comment]
You need to be aware of CVE-2016-10045 too - applying `escapeshellargs` to the sender address is not safe either. See [this advisory](https://legalhackers.com/advisories/PHPMailer-Exploit-Remote-Code-Exec-CVE-2016-10045-Vuln-Patch-Bypass.html). The only safe approach is to break RFC compliance for sender addresses.

Double quotes in real address are not unheard of, particularly with those that have apostrophes in their names, like `"Jane O'User"@example.com`, though they are definitely rare. RFC822 is a bit bonkers though - it even allows line breaks within addresses!

**Comment by cinu** [comment]
https://legalhackers.com/advisories/SwiftMailer-Exploit-Remote-Code-Exec-CVE-2016-10074-Vuln.html

**Comment by JHGitty** [comment]
PHPMailer got a second fix a few hours later and is released inside PHPMailer v5.2.21.
https://github.com/PHPMailer/PHPMailer/commit/833c35fe39715c3d01934508987e97af1fbc1ba0

**Comment by ponasromas** [comment]
So? Any fix released?

**Comment by barryvdh** [comment]
So to be clear, is this correct?

 - Only affects MailTransport (and FailoverTransport with MailTransport as failover?)
 - Only affected when using ReturnPath, Send or From options based on user input.
 - When using `filter_var($email, FILTER_VALIDATE_EMAIL)`, attack vector is not possible? (https://3v4l.org/XFp5D)

So can't we just filter the email before setting the extra headers? Or check if it's safe for exec. I can see it's validated based on RFC 2822, but that doesn't help apparently. 

It would invalidate some 'valid'  emails, but only for the 'From' part and only when for user-input + mail transport.

**Comment by pizzeys** [comment]
Most of that is correct - but filter_var(x, FILTER_VALIDATE_EMAIL) doesn't prevent it. Disallowing quotes in your email addresses does, to the best of my knowledge, if that is acceptable in your use case.

**Comment by pizzeys** [comment]
Just to clarify because I notice you used my example there - it's actually:

`->setFrom(array('"x@y.com\" -X/tmp/test"@test.php' => 'John Doe'))`

Didn't realise it would get mangled without code tags.

**Comment by barryvdh** [comment]
So based on the phpmailer fix, would this suffice?
https://github.com/swiftmailer/swiftmailer/pull/846

**Comment by barryvdh** [comment]
@pizzeys Indeed, your example still isn't valid according to filter_var, but I don't know all possible attack vectors.

**Comment by fabpot** [comment]
Fixed in 5.4.5. See https://github.com/swiftmailer/swiftmailer/blob/5.x/CHANGES#L10
